src/Symfony/Component/Config/Definition/EnumNode.php

```diff
@@ -78,13 +104,58 @@ protected function finalizeValue(mixed $value): mixed
     {
         $value = parent::finalizeValue($value);
 
-        if (!\in_array($value, $this->values, true)) {
-            $ex = new InvalidConfigurationException(\sprintf('The value %s is not allowed for path "%s". Permissible values: %s', json_encode($value), $this->getPath(), $this->getPermissibleValues(', ')));
-            $ex->setPath($this->getPath());
+        if (!$this->enumFqcn) {
+            if (!\in_array($value, $this->values, true)) {
+                throw $this->createInvalidValueException($value);
+            }
 
-            throw $ex;
+            return $value;
         }
 
-        return $value;
+        if ($value instanceof $this->enumFqcn) {
+            return $value;
+        }
+
+        if (!is_subclass_of($this->enumFqcn, \BackedEnum::class)) {
+            // value is not an instance of the enum, and the enum is not
+            // backed, meaning no cast is possible
+            throw $this->createInvalidValueException($value);
+        }
+
+        if (\is_string($value) || \is_int($value)) {
+            try {
+                $case = $this->enumFqcn::tryFrom($value);
```

Comment:
let's go with ::from() and catch ValueError|TypeError
and remove the check for null below